src/plugins/intel_cpu/src/graph.cpp

```diff
@@ -955,20 +914,124 @@ void Graph::Allocate(const std::vector<size_t>& syncNodesInds) {
+    for (const auto& edge : edges) {
+        if (edge->getStatus() == Edge::Status::Uninitialized) {
+            if (edge->getParent()->getParentEdges().empty() &&
+                one_of(edge->getParent()->getType(), Type::MemoryInput) &&
+                edge->inPlace(Edge::LOOK_UP)) {
+                edge->getParent()->resolveInPlaceEdges(Edge::LOOK_UP);
+            } else if (edge->getChild()->getChildEdges().empty() &&
+                one_of(edge->getChild()->getType(), Type::MemoryOutput) &&
+                edge->inPlace(Edge::LOOK_DOWN)) {
+                edge->getChild()->resolveInPlaceEdges(Edge::LOOK_DOWN);
+            }
+        }
+    }
+}
+
+int Graph::RegisterToAllocationContext(int offset, AllocationContext& context) {
+    auto syncNodesInds = CreateExecutionGraph();
+
+    ResolveInOutInPlaceEdges(graphEdges);
+
+    for (size_t i = 0, j = 0; i < graphNodes.size(); i++) {
+        const auto& node = graphNodes[i];
+        const auto inputExecIndex = i + offset;
+        offset = node->registerToAllocationContext(offset, context) - 1;
+        const auto outputExecIndex = i + offset;
+        context.execIndex[node] = {inputExecIndex, outputExecIndex};
+
+        if (j < syncNodesInds.size() && syncNodesInds[j] == i) {
+            context.syncPoints.push_back(inputExecIndex);
+            j++;
+        }
+    }
+
+    context.edges.insert(context.edges.end(), graphEdges.begin(), graphEdges.end());
+
+    return offset;
+}
+
+AllocationContext Graph::CreateAllocationContext(bool global) {
+    AllocationContext allocationContext;
+
+    if (global) {
+        RegisterToAllocationContext(0, allocationContext);
+    } else { // local allocation context. Used for the nodes with inner graph which are not updated yet
+        ResolveInOutInPlaceEdgesLegacy(graphEdges);
+
+        auto syncNodesInds = CreateExecutionGraph();
+
+        for (size_t i = 0; i < graphNodes.size(); i++) {
+            const auto& node = graphNodes[i];
+            const int inputExecIndex = i;
+            const int outputExecIndex = i;
+            allocationContext.execIndex[node] = {inputExecIndex, outputExecIndex};
+        }
+
+        allocationContext.edges = graphEdges;
+        allocationContext.syncPoints = syncNodesInds;
+    }
 
-    // resolve edges. Define which will be a view on others
-    //   NeedAllocation - real blob
-    //   NotAllocated - view on other blob, peer or in-place
-    for (auto& edge : graphEdges) edge->init();
+    return allocationContext;
+}
+
+void Graph::Allocate(bool globalAllocation) {
+    if (std::getenv("FORCE_LOCAL"))
+        globalAllocation = false;
+    // Set up the memory control subsystem.
+    auto memoryControl = globalAllocation ? m_context->getMemoryControl() : m_context->getNetworkMemoryControl()->createMemoryControlUnit();
+
+    if (memoryControl->allocated()) {
+        // std::cout << "Memory is already allocated for a subgraph: " << _name << "\n";
+        return;
+    }
+
```

Comment:
This is the way to keep memory allocation procedure generic across all the graphs.
So, no graphs are unique (i.e. outer graph or subgraphs)
The idea is that the first graph which is being "Activated" is responsible to allocate the memory for all the "context" it has, which includes all the subgraphs.
Basically, it will always be the outer graph which actually does this.
I am not 100% satisfied with this approach to be honest, but on the other hand it does make sense.